File: dashboard/graficos.js
/* Helpers de gráfico del dashboard.
   Mismas formas y reglas que src/graficos.py: barras solo donde el cero es
   significativo, puntos para las notas, grilla recesiva, etiqueta directa y
   forma además de color.

   Cada gráfico se dibuja al ANCHO REAL de su contenedor. Antes el SVG tenía un
   viewBox fijo y el navegador lo escalaba: en una tarjeta ancha el texto de los
   ejes quedaba el doble de grande que el de la página. */

const NS = 'http://www.w3.org/2000/svg';
const C = {
  azul: '#2a78d6', naranja: '#eb6834', verde: '#0ca30c', rojo: '#d03b3b', ambar: '#eda100',
  gris: '#898781',        // gris de MARCA (puntos "resto"); el texto usa --tenue
  ctx: '#c3c2b7', fondo: '#fcfcfb',
  seq: ['#cde2fb', '#9ec5f4', '#6da7ec', '#3987e5', '#2a78d6', '#1c5cab', '#104281']
};
const COLOR_TIPO = { 'Película': C.azul, 'Serie': C.naranja };

/* Convención chilena: coma decimal y punto de miles. */
const num = (v, d = 0) => v.toLocaleString('es-CL', { minimumFractionDigits: d, maximumFractionDigits: d });

const el = (t, attrs = {}) => {
  const n = document.createElementNS(NS, t);
  for (const k in attrs) n.setAttribute(k, attrs[k]);
  return n;
};
const texto = (x, y, s, cls, extra = {}) => { const n = el('text', { x, y, class: cls, ...extra }); n.textContent = s; return n; };
/* Marcas de eje en valores redondos (1-2-2,5-5 × 10^k). Un eje que marca
   6,65 · 6,79 · 6,94 obliga a leer cada número; uno que marca 6,7 · 6,8 · 6,9
   se entiende de un vistazo. */
function marcasEje(min, max, n = 5) {
  const bruto = (max - min) / n, pot = Math.pow(10, Math.floor(Math.log10(bruto)));
  const paso = [1, 2, 2.5, 5, 10].map(f => f * pot).find(p => p >= bruto);
  const out = [];
  for (let v = Math.ceil(min / paso - 1e-9) * paso; v <= max + 1e-9; v += paso) out.push(+v.toFixed(6));
  return { valores: out, dec: Math.max(0, -Math.floor(Math.log10(paso) + 1e-9)) + (paso / pot === 2.5 ? 1 : 0) };
}
const retraso = i => `animation-delay:${Math.min(i * 28, 400)}ms`;

/* ── Tooltip ──────────────────────────────────────────────────────────────── */
const tip = () => document.getElementById('tooltip');
function conTooltip(nodo, html) {
  const mover = e => {
    const t = tip(), r = t.getBoundingClientRect();
    let x = e.clientX + 14, y = e.clientY + 14;
    if (x + r.width > innerWidth - 8) x = e.clientX - r.width - 14;
    if (y + r.height > innerHeight - 8) y = e.clientY - r.height - 14;
    t.style.left = x + 'px'; t.style.top = y + 'px';
  };
  nodo.addEventListener('pointerenter', e => { const t = tip(); t.innerHTML = html; t.style.opacity = '1'; mover(e); });
  nodo.addEventListener('pointermove', mover);
  nodo.addEventListener('pointerleave', () => { tip().style.opacity = '0'; });
}

function lienzo(cont, alto) {
  const ancho = Math.max(Math.floor(cont.clientWidth), 300);
  const svg = el('svg', { width: ancho, height: alto, viewBox: `0 0 ${ancho} ${alto}`, role: 'img' });
  return [svg, ancho];
}
function vaciar(cont, filas, mensaje) {
  cont.innerHTML = '';
  if (!filas || !filas.length) {
    const d = document.createElement('div');
    d.className = 'vacio';
    d.textContent = mensaje || 'Sin títulos que cumplan los filtros seleccionados.';
    cont.appendChild(d);
    return true;
  }
  return false;
}

/* Formas: verde y rojo miden ΔE 4,1 bajo deuteranopía, así que cada segmento
   lleva además su propia forma. */
function marca(forma, cx, cy, r, color, i = 0) {
  const comun = { fill: color, stroke: C.fondo, 'stroke-width': 1.5, class: 'aparece', style: retraso(i) };
  if (forma === 'v') { const k = r * 1.25; return el('polygon', { points: `${cx - k},${cy - k * .8} ${cx + k},${cy - k * .8} ${cx},${cy + k}`, ...comun }); }
  if (forma === 's') return el('rect', { x: cx - r, y: cy - r, width: r * 2, height: r * 2, rx: 1.5, ...comun });
  if (forma === 'D') { const k = r * 1.2; return el('polygon', { points: `${cx},${cy - k} ${cx + k},${cy} ${cx},${cy + k} ${cx - k},${cy}`, ...comun }); }
  return el('circle', { cx, cy, r, ...comun });
}

/* ── Barras horizontales desde cero (solo conteos) ────────────────────────── */
function barrasH(cont, datos, { color = C.azul, formato = v => num(v), unidad = '' } = {}) {
  if (vaciar(cont, datos)) return;
  const fila = 27, izq = Math.min(150, Math.max(96, ...datos.map(d => d.etiqueta.length * 6.6 + 14)));
  const [svg, ancho] = lienzo(cont, datos.length * fila + 4);
  const der = 62, max = Math.max(...datos.map(d => d.valor)) || 1;
  datos.forEach((d, i) => {
    const y = i * fila + 3, w = Math.max((d.valor / max) * (ancho - izq - der), 2);
    const g = el('g', { class: 'marca-g' });
    g.appendChild(el('rect', { x: izq, y, width: w, height: fila - 10, fill: d.color || color, rx: 4, class: 'crece-x', style: retraso(i) }));
    g.appendChild(texto(izq - 10, y + 12.5, d.etiqueta, 'cat-texto', { 'text-anchor': 'end' }));
    g.appendChild(texto(izq + w + 8, y + 12.5, formato(d.valor), 'valor-texto'));
    conTooltip(g, `<b>${d.etiqueta}</b><br>${formato(d.valor)} ${unidad}`);
    svg.appendChild(g);
  });
  cont.appendChild(svg);
}

/* ── Dot plot (notas): la posición no necesita origen en cero ─────────────── */
function puntos(cont, datos, { formato = v => num(v, 2), resaltar = {} } = {}) {
  if (vaciar(cont, datos)) return;
  const fila = 26, izq = Math.min(190, Math.max(110, ...datos.map(d => d.etiqueta.length * 6.6 + 16))), der = 58;
  const [svg, ancho] = lienzo(cont, datos.length * fila + 30);
  let min = Math.min(...datos.map(d => d.valor)), max = Math.max(...datos.map(d => d.valor));
  const pad = (max - min) * .15 || .25; min -= pad; max += pad;
  const x = v => izq + ((v - min) / (max - min)) * (ancho - izq - der);
  const mp = marcasEje(min, max, 6);
  mp.valores.forEach(v => {
    svg.appendChild(el('line', { x1: x(v), x2: x(v), y1: 0, y2: datos.length * fila, class: 'grilla-linea' }));
    svg.appendChild(texto(x(v), datos.length * fila + 17, num(v, mp.dec), 'eje-texto', { 'text-anchor': 'middle' }));
  });
  datos.forEach((d, i) => {
    const y = i * fila + 13, cx = x(d.valor), col = resaltar[d.etiqueta] || d.color || C.azul;
    const g = el('g', { class: 'marca-g' });
    g.appendChild(el('line', { x1: izq, x2: cx, y1: y, y2: y, stroke: '#e8e7e1', 'stroke-width': 1.3, class: 'crece-x', style: retraso(i) }));
    g.appendChild(marca(d.forma || 'o', cx, y, 6, col, i));
    g.appendChild(texto(izq - 10, y + 4, d.etiqueta, 'cat-texto', { 'text-anchor': 'end' }));
    g.appendChild(texto(cx + 12, y + 4, formato(d.valor), 'valor-texto'));
    conTooltip(g, d.tip || `<b>${d.etiqueta}</b><br>${formato(d.valor)}`);
    svg.appendChild(g);
  });
  cont.appendChild(svg);
}

/* ── Barras verticales desde cero, con línea de referencia ────────────────── */
function barrasV(cont, datos, { referencia = null, etiquetaRef = '', formato = v => num(v, 2) + '×' } = {}) {
  if (vaciar(cont, datos)) return;
  const alto = 272, m = { arriba: 24, abajo: 46, izq: 44, der: 12 };
  const [svg, ancho] = lienzo(cont, alto);
  const mp = marcasEje(0, Math.max(...datos.map(d => d.valor), referencia || 0) * 1.12, 4);
  const max = mp.valores.at(-1) < Math.max(...datos.map(d => d.valor)) * 1.1 ? Math.max(...datos.map(d => d.valor)) * 1.15 : mp.valores.at(-1);
  const y = v => alto - m.abajo - (v / max) * (alto - m.arriba - m.abajo);
  const paso = (ancho - m.izq - m.der) / datos.length;
  mp.valores.filter(v => v <= max).forEach(v => {
    svg.appendChild(el('line', { x1: m.izq, x2: ancho - m.der, y1: y(v), y2: y(v), class: 'grilla-linea' }));
    svg.appendChild(texto(m.izq - 8, y(v) + 4, num(v, mp.dec), 'eje-texto', { 'text-anchor': 'end' }));
  });
  /* La línea de equilibrio va encima de las barras para que se vea cruzarlas, y
     su rótulo vive en la leyenda: dentro del gráfico chocaba con la barra más alta. */
  const capa = el('g');
  datos.forEach((d, i) => {
    const cx = m.izq + paso * i + paso / 2, w = Math.min(paso * .58, 70);
    const g = el('g', { class: 'marca-g' });
    g.appendChild(el('rect', { x: cx - w / 2, y: y(d.valor), width: w, height: Math.max(alto - m.abajo - y(d.valor), 1),
      fill: d.color || C.azul, rx: 4, class: 'crece-y', style: retraso(i * 2) }));
    g.appendChild(texto(cx, y(d.valor) - 8, formato(d.valor), 'valor-texto', { 'text-anchor': 'middle' }));
    g.appendChild(texto(cx, alto - m.abajo + 17, d.etiqueta, 'eje-texto', { 'text-anchor': 'middle' }));
    g.appendChild(texto(cx, alto - m.abajo + 32, 'n = ' + num(d.n), 'ctx-texto', { 'text-anchor': 'middle' }));
    conTooltip(g, `<b>Nota ${d.etiqueta}</b><br>La película típica recauda ${formato(d.valor)} lo invertido<br>${num(d.n)} películas`);
    capa.appendChild(g);
  });
  svg.appendChild(capa);
  if (referencia != null)
    svg.appendChild(el('line', { x1: m.izq, x2: ancho - m.der, y1: y(referencia), y2: y(referencia),
      stroke: C.rojo, 'stroke-width': 1.5, 'stroke-dasharray': '5 4', 'pointer-events': 'none' }));
  cont.appendChild(svg);
}

/* ── Dispersión con cuadrantes ────────────────────────────────────────────── */
function dispersion(cont, datos, o) {
  if (vaciar(cont, datos)) return;
  const alto = 430, m = { arriba: 40, abajo: 50, izq: 58, der: 24 };
  const [svg, ancho] = lienzo(cont, alto);
  const tx = o.log ? Math.log10 : (v => v);
  const rango = (a, f) => { const mn = Math.min(...a), mx = Math.max(...a), p = (mx - mn) * f || 1; return [mn - p, mx + p]; };
  const [x0, x1] = rango(datos.map(d => tx(d.x)), .12), [y0, y1] = rango(datos.map(d => d.y), .2);
  const X = v => m.izq + ((tx(v) - x0) / (x1 - x0)) * (ancho - m.izq - m.der);
  const Y = v => alto - m.abajo - ((v - y0) / (y1 - y0)) * (alto - m.arriba - m.abajo);
  const my = marcasEje(y0, y1, 5);
  my.valores.forEach(v => {
    svg.appendChild(el('line', { x1: m.izq, x2: ancho - m.der, y1: Y(v), y2: Y(v), class: 'grilla-linea' }));
    svg.appendChild(texto(m.izq - 8, Y(v) + 4, num(v, my.dec), 'eje-texto', { 'text-anchor': 'end' }));
  });
  [300, 500, 1000, 2000, 3000, 5000, 10000].filter(v => tx(v) > x0 && tx(v) < x1).forEach(v =>
    svg.appendChild(texto(X(v), alto - m.abajo + 16, num(v), 'eje-texto', { 'text-anchor': 'middle' })));
  const corte = { stroke: C.ctx, 'stroke-width': 1, 'stroke-dasharray': '5 4' };
  svg.appendChild(el('line', { x1: X(o.cortX), x2: X(o.cortX), y1: m.arriba - 10, y2: alto - m.abajo, ...corte }));
  svg.appendChild(el('line', { x1: m.izq, x2: ancho - m.der, y1: Y(o.cortY), y2: Y(o.cortY), ...corte }));
  (o.cuadrantes || []).forEach(q => {
    const px = q.ha === 'end' ? ancho - m.der : m.izq + 6, py = q.arriba ? m.arriba - 14 : alto - m.abajo - 22;
    svg.appendChild(texto(px, py, q.titulo, 'ctx-texto', { 'text-anchor': q.ha, style: `fill:${q.color};font-weight:700;letter-spacing:.06em` }));
    svg.appendChild(texto(px, py + 14, q.sub, 'ctx-texto', { 'text-anchor': q.ha }));
  });
  svg.appendChild(texto((m.izq + ancho - m.der) / 2, alto - 8, o.rotuloX, 'eje-texto', { 'text-anchor': 'middle' }));
  svg.appendChild(texto(-(alto - m.abajo + m.arriba) / 2, 14, o.rotuloY, 'eje-texto', { 'text-anchor': 'middle', transform: 'rotate(-90)' }));
  datos.forEach((d, i) => {
    const g = el('g', { class: 'marca-g' }), r = d.r || 6;
    g.appendChild(marca(d.forma || 'o', X(d.x), Y(d.y), r, d.color, i));
    if (d.etiqueta) g.appendChild(texto(X(d.x), Y(d.y) - r - 7, d.etiqueta, d.destaca ? 'valor-texto' : 'ctx-texto', { 'text-anchor': 'middle' }));
    conTooltip(g, d.tip);
    svg.appendChild(g);
  });
  cont.appendChild(svg);
}

/* ── Mancuerna (película contra serie) ────────────────────────────────────── */
function mancuerna(cont, datos, { formato = v => num(v, 2) } = {}) {
  if (vaciar(cont, datos, 'La comparación necesita películas y series: quita el filtro de tipo para verla.')) return;
  const fila = 30, izq = 104, der = 24;
  const [svg, ancho] = lienzo(cont, datos.length * fila + 32);
  let min = Math.min(...datos.flatMap(d => [d.a, d.b])), max = Math.max(...datos.flatMap(d => [d.a, d.b]));
  const pad = (max - min) * .16 || .3; min -= pad; max += pad;
  const X = v => izq + ((v - min) / (max - min)) * (ancho - izq - der);
  const mp = marcasEje(min, max, 6);
  mp.valores.forEach(v => {
    svg.appendChild(el('line', { x1: X(v), x2: X(v), y1: 0, y2: datos.length * fila, class: 'grilla-linea' }));
    svg.appendChild(texto(X(v), datos.length * fila + 17, num(v, mp.dec), 'eje-texto', { 'text-anchor': 'middle' }));
  });
  datos.forEach((d, i) => {
    const y = i * fila + 16, g = el('g', { class: 'marca-g' });
    g.appendChild(el('line', { x1: X(d.a), x2: X(d.b), y1: y, y2: y, stroke: C.ctx, 'stroke-width': 2.5, 'stroke-linecap': 'round', class: 'crece-x', style: retraso(i) }));
    g.appendChild(marca('o', X(d.a), y, 6, COLOR_TIPO['Película'], i));
    g.appendChild(marca('s', X(d.b), y, 5.5, COLOR_TIPO['Serie'], i));
    g.appendChild(texto(izq - 10, y + 4, d.etiqueta, 'cat-texto', { 'text-anchor': 'end' }));
    const dif = d.b - d.a;
    g.appendChild(texto((X(d.a) + X(d.b)) / 2, y - 9, (dif >= 0 ? '+' : '−') + num(Math.abs(dif), 2), 'ctx-texto', { 'text-anchor': 'middle' }));
    conTooltip(g, `<b>${d.etiqueta}</b><br>Película ${formato(d.a)} · ${num(d.na)} títulos<br>Serie ${formato(d.b)} · ${num(d.nb)} títulos`);
    svg.appendChild(g);
  });
  cont.appendChild(svg);
}

File: src/graficos.py
"""
Sistema visual del proyecto: paleta, tipografía y helpers de gráfico.

Un solo lugar define cómo se ve todo. Los notebooks no fijan colores ni tamaños
a mano: importan de aquí. Los mismos valores se replican en el dashboard, de
modo que una categoría conserva su color en el informe, en la presentación y en
Power BI (evidencia de IE5 y de IE9).

Los criterios que hay detrás de cada decisión están en `docs/decisiones_diseno.md`.

Reglas que este módulo hace cumplir por construcción:
  - Sin 3D, sin ejes truncados, sin doble eje, sin torta de más de 5 categorías.
  - El color nunca es el único canal: todo gráfico lleva etiqueta directa.
  - Grillas y ejes recesivos; el dato es lo único con contraste alto.
  - El título se redacta como conclusión, no como descripción.
"""

from __future__ import annotations

from pathlib import Path

import matplotlib as mpl
import matplotlib.pyplot as plt

RAIZ = Path(__file__).resolve().parent.parent
DIR_EXPLORATORIO = RAIZ / "images" / "exploratorio"
DIR_FINALES = RAIZ / "images" / "finales"

# ──────────────────────────────────────────────────────────────────────────────
# Paleta
#
# Paleta categórica validada: banda de luminosidad, piso de croma, separación
# bajo daltonismo (protanopía, deuteranopía, tritanopía) y contraste contra el
# fondo. El peor par adyacente da ΔE 9,1 bajo protanopía y 19,6 en visión normal,
# sobre un objetivo de 8 y un piso de 15 respectivamente.
#
# El orden de los slots es el mecanismo de seguridad, no decoración: se asignan
# SIEMPRE en este orden y nunca se ciclan. Un noveno color no se inventa — la
# cola se agrupa en "Otros".
# ──────────────────────────────────────────────────────────────────────────────

CATEGORICA = [
    "#2a78d6",  # 1 azul
    "#eb6834",  # 2 naranja
    "#1baf7a",  # 3 verde agua
    "#eda100",  # 4 amarillo
    "#e87ba4",  # 5 magenta
    "#008300",  # 6 verde
    "#4a3aa7",  # 7 violeta
    "#e34948",  # 8 rojo
]

# Color semántico fijo: Película siempre azul, Serie siempre naranja, en todos
# los visuales del proyecto. No se reasigna aunque cambie el orden de las barras.
COLOR_TIPO = {"Película": CATEGORICA[0], "Serie": CATEGORICA[1]}

# Rampa secuencial de un solo tono (claro → oscuro) para magnitud continua.
SECUENCIAL = ["#cde2fb", "#9ec5f4", "#6da7ec", "#3987e5", "#2a78d6", "#1c5cab", "#104281"]

# Acentos de lectura para los gráficos de cuadrantes (visuales 19 y 21).
# Son colores de ESTADO (bueno / malo), no "un color más" de una serie, y por eso
# no entran nunca en la rotación categórica.
ACENTO = {
    "oportunidad": "#0ca30c",  # poco volumen, buena nota → invertir
    "sobreinvertido": "#d03b3b",  # mucho volumen, nota baja → revisar
    "neutro": "#898781",  # el resto, en gris: contexto, no protagonista
}

# El verde y el rojo de arriba miden ΔE 4,1 bajo deuteranopía: para un lector con
# esa condición son el mismo color. Es el resultado del validador, no una
# sospecha. Como el par verde/rojo comunica "bueno / malo" mejor que cualquier
# alternativa que sí pase, se conserva el color Y se agrega un segundo canal:
# cada segmento tiene su propia FORMA de marca. Quien no distinga los tonos
# sigue leyendo el gráfico por la forma, por la posición en el cuadrante y por
# la etiqueta directa.
MARCA = {
    "oportunidad": "o",  # círculo
    "priorizar": "o",
    "sobreinvertido": "v",  # triángulo hacia abajo
    "revisar": "v",
    "prestigio": "D",  # rombo
    "eficiencia": "s",  # cuadrado
    "neutro": "o",
}

# Tinta y cromo del gráfico. El dato lleva el contraste; el andamiaje se apaga.
FONDO = "#fcfcfb"
TINTA = "#0b0b0b"
TINTA_SECUNDARIA = "#52514e"
# Gris de texto secundario. Era #898781, que da 3,5:1 sobre el fondo: bajo el
# 4,5:1 que WCAG AA exige para texto chico, y justo se usaba en ejes, notas y
# pies. #6e6d68 da 5,05:1. El gris claro sigue existiendo, pero solo para
# MARCAS (ACENTO["neutro"]), donde el mínimo para objetos gráficos es 3:1.
TINTA_TENUE = "#6e6d68"
GRILLA = "#e1e0d9"
EJE = "#c3c2b7"

# Gris de de-énfasis: el resto de las categorías cuando una es el punto.
GRIS_CONTEXTO = "#c3c2b7"

# Inter: diseñada para pantallas y cifras, con dígitos tabulares y una x alta
# que se lee bien en tamaños chicos. Es la misma familia del dashboard y del
# informe. DejaVu queda de respaldo si Inter no está instalada.
TIPOGRAFIA = ["Inter", "DejaVu Sans", "Liberation Sans", "Helvetica", "Arial", "sans-serif"]


def aplicar_estilo() -> None:
    """
    Fija el estilo del proyecto en matplotlib. Llamar una vez por notebook.

    Todo lo que apaga andamiaje (bordes, grilla, ticks) está aquí y no repetido
    gráfico a gráfico: así ningún visual se sale del sistema por descuido.
    """
    mpl.rcParams.update(
        {
            "figure.facecolor": FONDO,
            "figure.dpi": 110,
            "savefig.dpi": 200,
            "savefig.facecolor": FONDO,
            "savefig.bbox": "tight",
            "savefig.pad_inches": 0.3,
            "font.family": "sans-serif",
            "font.sans-serif": TIPOGRAFIA,
            "font.size": 10,
            "text.color": TINTA,
            "axes.facecolor": FONDO,
            "axes.edgecolor": EJE,
            "axes.labelcolor": TINTA_SECUNDARIA,
            "axes.labelsize": 10,
            "axes.titlesize": 13,
            "axes.titleweight": "bold",
            "axes.titlecolor": TINTA,
            "axes.spines.top": False,
            "axes.spines.right": False,
            "axes.spines.left": False,
            "axes.grid": True,
            "axes.axisbelow": True,
            "grid.color": GRILLA,
            "grid.linewidth": 0.8,
            "xtick.color": TINTA_TENUE,
            "ytick.color": TINTA_SECUNDARIA,
            "xtick.labelsize": 9,
            "ytick.labelsize": 9.5,
            "xtick.bottom": False,
            "ytick.left": False,
            "legend.frameon": False,
            "legend.fontsize": 9.5,
            "lines.linewidth": 2,
            "lines.markersize": 8,
        }
    )


def num_es(valor, decimales: int = 0) -> str:
    """
    Formatea un número a la convención chilena: coma decimal y punto de miles.

    Un eje que dice `6.5` se lee como "seis mil quinientos" en el idioma de la
    audiencia. Es un detalle, pero es exactamente el tipo de fricción que el
    lector paga con atención.
    """
    texto = f"{valor:,.{decimales}f}"
    return texto.replace(",", "\x00").replace(".", ",").replace("\x00", ".")


def eje_es(ax, eje: str = "x", decimales: int = 1) -> None:
    """Aplica la convención numérica en español a las marcas de un eje."""
    from matplotlib.ticker import FuncFormatter

    formateador = FuncFormatter(lambda v, _: num_es(v, decimales))
    (ax.xaxis if eje == "x" else ax.yaxis).set_major_formatter(formateador)


def titular(ax, conclusion: str, subtitulo: str | None = None, aire: int = 0) -> None:
    """
    Escribe el título como conclusión y el subtítulo como alcance.

    El título dice lo que el lector debe concluir ("Drama lidera en volumen pero
    no en nota"), no lo que el gráfico contiene ("Géneros por cantidad"). El
    subtítulo lleva el alcance: qué subconjunto, qué umbral, qué unidad. Es la
    manera de que un gráfico se explique solo cuando el Gerente lo mira sin
    nosotros al lado.
    """
    ax.set_title(conclusion, loc="left", pad=(26 if subtitulo else 14) + aire)
    if subtitulo:
        ax.annotate(
            subtitulo,
            xy=(0, 1),
            xycoords="axes fraction",
            xytext=(0, 12),
            textcoords="offset points",
            ha="left",
            va="bottom",
            fontsize=9.5,
            color=TINTA_SECUNDARIA,
        )


def pie_de_fuente(ax, texto: str = "", desplazamiento: int = -42) -> None:
    """Nota al pie con la fuente y el alcance declarado. Va en todos los visuales."""
    base = "Fuente: catálogo Netflix 2010–2025 (31.991 títulos). Datos de catálogo, no de usuarios."
    ax.annotate(
        f"{base} {texto}".strip(),
        xy=(0, 0),
        xycoords="axes fraction",
        xytext=(0, desplazamiento),
        textcoords="offset points",
        ha="left",
        va="top",
        fontsize=8,
        color=TINTA_TENUE,
    )


def colores_enfasis(etiquetas, destacadas: dict[str, str]) -> list[str]:
    """
    Pinta solo lo que es el punto del gráfico y manda el resto al gris de contexto.

    El énfasis es la forma más desaprovechada: cuando la historia es "estas dos
    categorías", usar ocho colores entierra el hallazgo en decoración.
    `destacadas` mapea etiqueta → color.
    """
    return [destacadas.get(e, GRIS_CONTEXTO) for e in etiquetas]


def etiquetar_barras(ax, barras, valores, formato="{:.0f}", dentro_desde=0.75) -> None:
    """
    Etiqueta directa sobre cada barra: el valor exacto sin que el ojo viaje al eje.

    Las barras largas llevan la etiqueta dentro en blanco y las cortas fuera en
    tinta, para que nunca quede texto ilegible sobre un fondo de color. Esto es
    también lo que da el segundo canal de lectura: el color deja de ser el único
    portador de información.
    """
    limite = max(valores) if len(valores) else 0
    for barra, valor in zip(barras, valores):
        ancho = barra.get_width()
        adentro = limite and (ancho / limite) >= dentro_desde
        ax.annotate(
            formato.format(valor),
            xy=(ancho, barra.get_y() + barra.get_height() / 2),
            xytext=(-6 if adentro else 6, 0),
            textcoords="offset points",
            ha="right" if adentro else "left",
            va="center",
            fontsize=9,
            fontweight="bold",
            color="#ffffff" if adentro else TINTA_SECUNDARIA,
        )


def puntos_ordenados(
    ax,
    etiquetas,
    valores,
    colores,
    formato="{:.2f}",
    marcas=None,
    etiqueta_derecha=None,
    titulo_derecha="",
) -> None:
    """
    Gráfico de puntos (dot plot) horizontal, ordenado por valor.

    **Por qué puntos y no barras.** La longitud de una barra codifica magnitud
    desde cero: si el eje no parte en cero, la barra miente — un 6,5 se ve el
    doble que un 6,2. Cuando el rango útil es estrecho (todas las notas viven
    entre 5,9 y 7,0), o se trunca el eje, que está prohibido, o las barras se
    vuelven indistinguibles.

    El punto codifica con **posición**, no con longitud, y la posición no
    necesita origen en cero: es legítimo acercar el eje al rango de los datos.
    Por eso todo ranking de notas del proyecto es un dot plot y no un gráfico
    de barras. La línea guía tenue solo conduce el ojo de la etiqueta al punto.

    `etiqueta_derecha` agrega una segunda columna de contexto (el volumen), de
    modo que el lector vea nota y cantidad sin cambiar de gráfico.

    `marcas` da a cada punto su propia forma. Es obligatorio cuando los colores
    vienen de `ACENTO`: verde y rojo son indistinguibles bajo deuteranopía y la
    forma es lo que sostiene la lectura en ese caso.
    """
    posiciones = range(len(etiquetas))
    minimo = min(valores)

    marcas = list(marcas) if marcas is not None else ["o"] * len(valores)

    for y, valor, color, marca in zip(posiciones, valores, colores, marcas):
        ax.plot([minimo, valor], [y, y], color=GRILLA, lw=1.2, zorder=1, solid_capstyle="round")
        ax.plot([valor], [y], marca, color=color, markersize=11, zorder=3,
                markeredgecolor=FONDO, markeredgewidth=1.5)
        ax.annotate(formato(valor) if callable(formato) else formato.format(valor),
                    xy=(valor, y), xytext=(13, 0),
                    textcoords="offset points", ha="left", va="center",
                    fontsize=9.5, fontweight="bold", color=TINTA)

    ax.set_yticks(list(posiciones), list(etiquetas))
    ax.set_ylim(-0.7, len(etiquetas) - 0.3)

    if etiqueta_derecha is not None:
        for y, texto in zip(posiciones, etiqueta_derecha):
            ax.annotate(texto, xy=(1, y), xycoords=("axes fraction", "data"),
                        xytext=(12, 0), textcoords="offset points",
                        ha="right", va="center", fontsize=8.5, color=TINTA_TENUE)
        ax.annotate(titulo_derecha, xy=(1, 1), xycoords="axes fraction",
                    xytext=(12, 8), textcoords="offset points", ha="right", va="bottom",
                    fontsize=8.5, fontweight="bold", color=TINTA_TENUE)


def leyenda_segmentos(ax, entradas: dict[str, str], **kwargs) -> None:
    """
    Leyenda con marcador de color y texto: el color nunca viaja solo.

    Con `abajo=True` se ubica bajo el área de datos, para no tapar marcas.
    `marcadores` permite que el símbolo de la leyenda sea el mismo que el del
    gráfico (círculo, cuadrado): si no coinciden, la leyenda deja de ser leyenda.
    Es obligatorio pasarlo cuando el gráfico usa los colores de `ACENTO`, porque
    ahí la forma es la que carga la distinción para un lector con daltonismo.
    """
    from matplotlib.lines import Line2D

    marcadores = kwargs.pop("marcadores", {})
    handles = [
        Line2D([], [], marker=marcadores.get(texto, "o"), linestyle="", markersize=9,
               color=color, label=texto)
        for texto, color in entradas.items()
    ]
    kwargs.setdefault("handletextpad", 0.4)
    if kwargs.pop("abajo", False):
        # Anclada en PUNTOS bajo el eje, no en fracción de los ejes: la fracción
        # depende del alto de la figura, así que un mismo valor deja la leyenda
        # encima de las marcas en un gráfico y flotando en otro. En puntos, la
        # distancia al rótulo del eje es siempre la misma.
        from matplotlib.transforms import offset_copy

        puntos = kwargs.pop("altura", 46)
        transformacion = offset_copy(
            ax.transAxes, fig=ax.figure, y=-puntos / 72, units="inches"
        )
        kwargs.update(
            loc="upper left",
            bbox_to_anchor=(0, 0, 1, 0),
            bbox_transform=transformacion,
            mode="expand",
            borderaxespad=0,
        )
        kwargs.setdefault("ncol", len(entradas))
    else:
        kwargs.setdefault("loc", "lower right")
    ax.legend(handles=handles, labelcolor=TINTA_SECUNDARIA, **kwargs)


def limpiar_ejes(ax, eje_valor: str = "x") -> None:
    """Deja la grilla solo en el eje del valor. Una grilla cruzada es ruido."""
    ax.grid(axis=eje_valor, visible=True)
    ax.grid(axis="y" if eje_valor == "x" else "x", visible=False)
    ax.set_axisbelow(True)


def guardar(fig, nombre: str, carpeta: Path = DIR_EXPLORATORIO, mostrar: bool = True) -> Path:
    """
    Guarda el gráfico como PNG y, dentro de un notebook, lo muestra.

    Muestra el ARCHIVO guardado y no la figura en memoria: lo que se ve en el
    notebook es exactamente lo que va al informe. Antes la figura se cerraba sin
    mostrarse, y el notebook de análisis quedaba sin un solo gráfico visible.
    """
    carpeta.mkdir(parents=True, exist_ok=True)
    ruta = carpeta / (nombre if nombre.endswith(".png") else f"{nombre}.png")
    fig.savefig(ruta)
    plt.close(fig)
    if mostrar:
        try:
            from IPython import get_ipython
            from IPython.display import Image, display

            if get_ipython() is not None:
                display(Image(filename=str(ruta), width=920))
        except ImportError:
            pass
    return ruta
